Proposals to make configuration easier to use.

shawn-hurley opened the pull request:
**Describe what this PR does and why we need it**:
Proposals to determine how to we will change configuration values to make them easier to use.

jmrodri commented on docs/proposals/configuration_changes.md:
```diff
@@ -0,0 +1,71 @@
+Create a package for creating and managing configuration values as a value store. This will be used to remove the necessity of each package to be loaded just for configuration values. The other benefit is that this package could be a more robust solution to the configuration, Allowing us to do a better job of watching for updates to the configmap and managing subscriptions to those changes. This  ...
+
+### Work Items
+* Create a specific package for managing configuration 
```
I'm assuming the configuration file stays the same, since you didn't specify any changes in the document.

shawn-hurley replied:
Yes, that was my hope. I was hoping that we could keep yaml because I think it looks nice and fits with the config map stuff that is already in place.

jmrodri replied:
Like I said below, I think you're right, the current yaml would look nicer than my proposed yaml + properties file mixture.

jmrodri commented on docs/proposals/configuration_changes.md:
```diff
@@ -0,0 +1,71 @@
+            user:shurley
+        ]
+    ] 
+    dao:map[
```
This structure seems way more complicated since it's like the specific structs except as maps. Maybe I'm being myopic or simple minded but I always envisioned just a simple flat map.

```yaml
---
registry.docker.type: dockerhub
registry.docker.name: docker
registry.docker.url: https://registry.hub.docker.com
registry.docker.user: DOCKERHUB_USER
registry.docker.pass: DOCKERHUB_PASS
registry.docker.org: DOCKERHUB_ORG
registry.docker.fail_on_error: false
registry.rhcc.type: rhcc
registry.rhcc.name: rhcc
registry.rhcc.url: registry.access.stage.redhat.com
registry.rhcc.fail_on_error: true
registry.rhcc.white_list.1: "^legitimate.*-apb$"
registry.rhcc.black_list.2: "malicious.*-apb$"
registry.rhcc.black_list.3: "^specific-blacklist-apb$"
dao.etcd_host: localhost
dao.etcd_port: 2379
log.logfile: /var/log/ansible-service-broker/asb.log
log.stdout: true
log.level: debug
log.color: true
openshift.host: ""
openshift.ca_file: ""
openshift.bearer_token_file: ""
openshift.apb_sandbox_role: "admin"
openshift.image_pull_policy: IfNotPresent
broker.bootstrap_on_startup: true
broker.dev_broker: true
broker.launch_apb_on_bind: false
broker.recovery: true
broker.output_request: true
broker.ssl_cert_key: /path/to/key
broker.ssl_cert: /path/to/cert
broker.refresh_interval: "600s"
broker.auth.basic.type: basic
broker.auth.basic.enabled: true
```

Then the `Config` struct would look like this:

```golang
type Config interface {
    ToMap() map[string]string
    ContainsKey(key string) bool
    GetBoolean(string, bool) bool
    GetString(key string) string
    GetList(key string) []string
    SubsetMap(prefix string) map[string]string
     ... // many more utility methods
}

type MapConfiguration struct {
    configMap map[string]string
}

func NewConfigurationFromMap(configmap map[string]string) Config {
    return MapConfiguration{configMap: configmap}
}

func NewConfiguration() Config {
    return MapConfiguration{}
}

func (c MapConfiguration) SubsetMap(prefix string) map[string]string {
   // return a new map with all the entries that match the prefix from the embedded configMap
}
```

If I call `config.SubsetMap("registry.docker")` I will get a map with only the following keys:

```
registry.docker.type: dockerhub
registry.docker.name: docker
registry.docker.url: https://registry.hub.docker.com
registry.docker.user: DOCKERHUB_USER
registry.docker.pass: DOCKERHUB_PASS
registry.docker.org: DOCKERHUB_ORG
registry.docker.fail_on_error: false
```

If you called `config.GetBoolean("registry.docker.fail_on_error")` it will return false. Store everything as simple strings.

shawn-hurley replied:
:+1: to the interface 
I think that this would totally work, it is just that we would
1. have to transform the yaml map to this structure
2. For calls like this, we would need to loop through every key to build the new map and return that, while the map approach would be constant time to retrieve each part.
3. returning map[string]string will make the caller use the [strconv](https://golang.org/pkg/strconv/) package, where if it is of type interface, they can just cast the value. I am not opposed to it, just making sure we are all in agreement. 

I am not opposed to doing the extra work in the config package, I just don't know if it will be easier to work with.

jmrodri replied:
So after thinking about this overnight and looking at the yaml I proposed, I'm not sure I like it. The original one we had seems better. So -1 to my proposed flat file approach :)

As for map[string]string vs map[string]interface{}. I always struggle with configuration since they're usually just values in a file so treating them all as strings seems logical to me. I guess I fear if we use interface{} you could put in some custom struct in there is that really what we want?

Regarding the caller having to use strconv, I don't understand why? If we offer them conversion methods like `GetInt(key string) int`. For example, if I want the port number for etcd as an int I'd call `config.GetInt("dao.etcd_port")` and get back `2379`. The implementation of GetInt would have to do the conversions.  But if we used `interface{}` no one has to convert. hrm.

jmrodri commented on docs/proposals/configuration_changes.md:
```diff
@@ -0,0 +1,71 @@
+* Create a public API that will give the caller multiple ways to retrieve data.
+    - `MustGet` -> returns a single value and will panic if not able to retrieve the data
+    - `MustGet/Get<Type>` -> Allows for typed returns so the caller does not have to worry about type casting.
+    - Ability to specify a `map[string]interface{}` as a return value so one could say `config.MustGetMapConfig("broker")` and would retrieve the map of values for the broker configuration section. 
```
I'd like to see this a little more. I don't really care for the `Must`. +1 to typed getters:

```
GetInt(key) int
GetString(key) string
```

Instead of the `Must` something that returns the default value if not present:
`GetSafeInt(key string, defValue int) int`

eriknelson replied:
Is there an alternative flavor of getter, like a `MayGet`?

If not, I'd just name them `Get`. Instead of the caller providing the default, should that be behind the `Getter` so it's configured in one place (the config package)? Trying to avoid having gets spread out across packages and possibly have each of those calls potentially use different defaults. Someone could easily come in and change it, and not know about the other calls.

jmrodri commented:
Okay here's what I'd like to do:

- create a Config interface with a bunch of utility methods
- create a Config implementation of that interface `MapConfiguration` or whatever is appropriate.
- keep the same format we have, as you suggested
- I'd get rid of the `Must` stuff just use `Get`s and `Put`s if we want to be able to programmatically update the config at runtime (in memory of course)
-stick with your map of map (of map of map sorry hahaha) implementation. Since it is hidden behind the interface and struct we really won't see it.

Some of the requirements I'd like are:

- given a top level key get a submap from the config. i.e. given "broker" give me the broker. submap
- be able to get a key as an int, string or string slice.

I think that's pretty much what you were proposing, just a few requirements.

jmrodri commented:
@eriknelson any thoughts on this?

shawn-hurley commented:
@jmrodri Just for my sanity: 
> be able to get a key as an int, string or string slice.

you mean a config value from a key correct?

eriknelson commented on docs/proposals/configuration_changes.md:
```diff
@@ -0,0 +1,71 @@
+## Configuration Changes
+
+### Goal
+Create a package for creating and managing configuration values as a value store. This will be used to remove the necessity of each package to be loaded just for configuration values. The other benefit is that this package could be a more robust solution to the configuration, Allowing us to do a better job of watching for updates to the configmap and managing subscriptions to those changes. This  ...
```
Is there a use case for config changes changing while a broker is running? Right now we just update the broker-config and bounce the pod.

shawn-hurley replied:
I think that we should eventually do this. I think it is a nice to have feature down the road, but that we should eventually have it. I don't see why, when I update the config map I then have to bounce the pod as the ideal end state IMO. I am not saying that as part of the initial PR to move to the new configuration that needs to work, just mentioning that moving to this new way will enable this feature in the future.

Does that make any sense?

eriknelson replied:
So you want to edit the config file and have that trigger something in the broker that reconfigures it? I guess I just don't see this being used often, unless I'm misunderstanding.

rthallisey commented on docs/proposals/configuration_changes.md:
```diff
@@ -0,0 +1,71 @@
+
+### Goal
+Create a package for creating and managing configuration values as a value store. This will be used to remove the necessity of each package to be loaded just for configuration values. The other benefit is that this package could be a more robust solution to the configuration, Allowing us to do a better job of watching for updates to the configmap and managing subscriptions to those changes. This  ...
+
```
+1

rthallisey commented on docs/proposals/configuration_changes.md:
````diff
@@ -0,0 +1,71 @@
+m["broker"] = map["dev_broker": true...]
+m["registry"] = [map["name": "dh"...], map["name": "play"...]]
+```
+* Create a syntax, and document, for retrieving a single configuration value from the list. I.E `config.MustGetBoolConfig("broker.dev_broker")`
````
+1 to an API.

rthallisey commented on docs/proposals/configuration_changes.md:
````diff
@@ -0,0 +1,71 @@
+```
+* Create a syntax, and document, for retrieving a single configuration value from the list. I.E `config.MustGetBoolConfig("broker.dev_broker")`
+* Create a public API that will give the caller multiple ways to retrieve data.
+    - `MustGet` -> returns a single value and will panic if not able to retrieve the data
````
I'm not a fan of ```MustGet```.  I think a single API call makes more sense with the caller processing the possible return values and failures.

rthallisey commented on docs/proposals/configuration_changes.md:
```diff
@@ -0,0 +1,71 @@
+* Create a public API that will give the caller multiple ways to retrieve data.
+    - `MustGet` -> returns a single value and will panic if not able to retrieve the data
+    - `MustGet/Get<Type>` -> Allows for typed returns so the caller does not have to worry about type casting.
+    - Ability to specify a `map[string]interface{}` as a return value so one could say `config.MustGetMapConfig("broker")` and would retrieve the map of values for the broker configuration section. 
```
I like that we can retrieve sections of the config info. Do you have an example of how we can use this? If you don't it's not a big deal because this can come in handy down the road.  My thought is this could be useful when we handle upgrading the broker.

shawn-hurley commented:
@jmrodri @rthallisey 
I have added a comparison to a library that already exists, as @eriknelson has asked for. 

I think that we should have a discussion if the weight of another vendor dependence is worth the benefit of not having to write the config package. 

I am leaning towards saying it is worth the dependence because this is something that we have to do, but not something I want to spend time on. I think that the interface is almost exactly what we were looking for and is used by some big projects so we can make some assumption that it is not completely broken. 

Also as a side note,
If we still go with creating our own config package, I did update the interface, to only have `Get` and `Put`methods and removed the concept of a `MustGet` method.

jmrodri commented:
@shawn-hurley @eriknelson @rthallisey I don't mind writing our own Config struct. Every project I've been on seems to write one and they've all been different. On RHN/Satellite 5 our configs were a series of .conf files with a hierarchy, on Candlepin we had a properties file (name=value) which was just string=string. 

So unless we need all the features of `viper` I'd vote just writing our own `Config`.

To @rthallisey I believe shawn does have an example of being able to get an entire section. So I should be able to get a map of the entire "broker" config, etc.

eriknelson commented:
> I am leaning towards saying it is worth the dependence because this is something that we have to do, but not something I want to spend time on. I think that the interface is almost exactly what we were looking for and is used by some big projects so we can make some assumption that it is not completely broken.

+1, this was mostly my thinking. I also don't see the addition of a vendor lib like this as an appreciable drawback; is the complaint the size of extra files? Doesn't seem like a big deal IMO.

jmrodri commented:
@shawn-hurley @eriknelson it isn't really about the size of the vendor (though viper does have its own set of deps that will pull in). It's just that config isn't that hard and wouldn't be much code. And I've always had to modify it to suit our particular needs. But at this point, we've spent more time debating than it would've taken to write a first cut at the config. So do whatever is easier.

rthallisey commented:
I'd rather we make our own pkg. We control it, it's not much code, and it cleans things up.

shawn-hurley commented:
I have updated the proposal to include the decision around building our own. 

Once Travis is completed can someone merge?
@jmrodri @eriknelson @rthallisey one last check that we are on the same page
